Implement opening editor timestamp links

RatinFX opened the pull request:
Alongside a lot of Test cases and some Error Notifications.

That being said, I couldn't exactly replicate what Stable does in some weird/unusual cases without funny lines of code, though I'm not sure if we even care about these scenarios (taking a look at the original code would surely help in this case).

Some unusual case examples from Stable:
- (currently works) `00:00:000 (0)`, `(1)`, `(2)`, and so on always selects the first object
- `00:00:000 (1,2)` - even though you can only get a timestamp like this from the std editor - it selects the first 2 objects in both std and in mania editors

Every feedback is welcome, I tried my best to understand where things belong and place files/methods accordingly.

RatinFX committed:
Added OpenEditorTimestamp base implementation

RatinFX committed:
Added ExclamationTriangle Icon to notifications

RatinFX committed:
Renamed Editor method

RatinFX committed:
Added OpenEditorTimestamp Tests

RatinFX committed:
Fixed up tests

RatinFX committed:
Moved error checks into Editor
- Invoke Action on error to Notify user
- added some comments

RatinFX committed:
Ensure every SelectedItem is alive and has Blueprint

LittleEndu commented on osu.Game/Screens/Edit/EditorTimestampParser.cs:
```diff
@@ -0,0 +1,101 @@
+{
+    public static class EditorTimestampParser
+    {
+        private static readonly Regex timestamp_regex = new Regex(@"^(\d+:\d+:\d+)(?: \((\d+(?:[|,]\d+)*)\))?$", RegexOptions.Compiled);
```
A timestamp regex already exists in message formatter:

https://github.com/ppy/osu/blob/master/osu.Game/Online/Chat/MessageFormatter.cs#L44-L46

RatinFX replied:
Oh nice, can we make this public or do I just copy it?
I'd rather use the existing one just in case something changes in the future

peppy replied:
If anything, I'd probably move that one into this static class and reference it here instead.

RatinFX replied:
Moving `time_regex` in here, making it public, and referencing it in `MessageFormatter`, surely I understood correctly

bdach commented on osu.Game/Screens/Edit/EditorTimestampParser.cs:
```diff
@@ -0,0 +1,101 @@
+            return (times[0] * 60 + times[1]) * 1_000 + times[2];
+        }
+
+        public static List<HitObject> GetSelectedHitObjects(IReadOnlyList<HitObject> editorHitObjects, string objectsGroup, double position)
```
The inverse of this operation, i.e. generating the selection text, is part of ruleset API via:

https://github.com/ppy/osu/blob/bc9cdb4ce07fcc7fc16117121f6bb620a8094c36/osu.Game/Rulesets/Edit/HitObjectComposer.cs#L529

Examples:

https://github.com/ppy/osu/blob/bc9cdb4ce07fcc7fc16117121f6bb620a8094c36/osu.Game.Rulesets.Osu/Edit/OsuHitObjectComposer.cs#L106-L107
https://github.com/ppy/osu/blob/bc9cdb4ce07fcc7fc16117121f6bb620a8094c36/osu.Game.Rulesets.Mania/Edit/ManiaHitObjectComposer.cs#L51-L52

Due to this, I would much rather have the individual rulesets / their composers handle this parsing, too, otherwise ruleset authors have the capability of formatting selections in their own ways, but have to rely on the main game assembly actually being able to handle them.

RatinFX replied:
Right, I'm a bit lost on how could I access the current `HitObjectComposer` in either `Editor.cs` or `EditorTimestampParser.cs`

bdach replied:
I think that works fine, sure.

bdach commented on osu.Game/Screens/Edit/Editor.cs:
```diff
@@ -1137,6 +1138,62 @@ private void cancelExit()
             loader?.CancelPendingDifficultySwitch();
         }
 
+        public void SeekAndSelectHitObjects(string timestamp, Action<LocalisableString> onError)
```
Probably would just call it `HandleTimestamp` at this point. Too specific to be named anything else.

Also the `onError` thing shouldn't exist really - there's not much point in marshaling the `NotificationsOverlay` through `OsuGame`. `INotificationsOverlay` is already available in `Editor`; just use that directly.

RatinFX replied:
Renamed and moved Notifications inside

bdach commented on osu.Game/Screens/Edit/Editor.cs:
```diff
@@ -1137,6 +1138,62 @@ private void cancelExit()
+            // Only seeking is necessary
+            if (string.IsNullOrEmpty(objectsGroup))
+            {
+                if (clock.IsRunning)
+                    clock.Stop();
+
+                clock.Seek(position);
+                return;
```
To reduce code duplication, I'd seek first, early-return if the objects portion is empty, and handle selection at the very end.

Also, if we're unconditionally stopping clock, the seek should probably be smooth (`SeekSmoothlyTo()`)

RatinFX replied:
The reason for 2 Seek calls: whenever the objects portion is not empty, it moves to the next closest object instead of the given timestamp.

I can move the mentioned `nextObject` and `Seek` codes below position init to make them seem related and I'm not sure if we avoid code duplication as we would either need to:
- check `objectsGroup` twice - for position init and early return,
- or `Seek` twice - in early return and after setting the new position.

Sure we change `Seek()` to `SeekSmoothlyTo()`, though I like the snappy-Seek more when you click the link.

bdach commented on osu.Game/Screens/Edit/Compose/Components/EditorBlueprintContainer.cs:
```diff
@@ -144,6 +149,15 @@ protected override void SelectAll()
             SelectedItems.AddRange(Beatmap.HitObjects.Except(SelectedItems).ToArray());
         }
 
+        protected void SetHitObjectsAlive(object sender, NotifyCollectionChangedEventArgs e)
```
- doesn't need to be `protected`
- probably needs better name (`KeepHitObjectsAlive`), I'd drop the `sender` arg too

RatinFX replied:
1. `private` instead i assume?
2. renamed
3. wouldn't dropping `sender` here cause problems in event un/subscription or do I not need to add this to `Dispose()` to begin with? (`Beatmap.SelectedHitObjects.CollectionChanged += / -=` at line 53 & 183)

bdach replied:
> 1. `private` instead i assume?

Correct.

> 3. wouldn't dropping `sender` here cause problems in event un/subscription or do I not need to add this to `Dispose()` to begin with? (`Beatmap.SelectedHitObjects.CollectionChanged += / -=` at line 53 & 183)

Fair point, can keep as is, but in that case the method should just be named `selectionChanged` or something.

bdach commented on osu.Game.Tests/Visual/Editing/TestSceneOpenEditorTimestamp.cs:
```diff
@@ -0,0 +1,328 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable disable
```
New code shouldn't use `#nullable disable`

RatinFX replied:
Removed

bdach commented on osu.Game.Tests/Visual/Editing/TestSceneOpenEditorTimestamp.cs:
```diff
@@ -0,0 +1,328 @@
+        }
+
+        [Test]
+        public void TestSelectionForOsu()
```
In line with the other comment, these selection tests should probably go into the ruleset test projects.

RatinFX replied:
Separated Test cases o7

RatinFX committed:
Removed nullable line from Test

RatinFX committed:
Renamed method and moved Notifications inside

RatinFX committed:
Renamed method and made private

RatinFX committed:
HitObject Selection logic and separation for gamemodes
+ moved time_regex into EditorTimestampParser

RatinFX committed:
Renamed CollectionChanged event handler

RatinFX committed:
Forgot a clock.Stop call

RatinFX committed:
Separate Test cases by relevant rulesets

RatinFX commented:
I uploaded everything new.

- Added HitObject selection parser that can be overwritten for every game mode: https://github.com/RatinFX/osu/blob/81caa854e6a7ba3251658f1cd174bf6bad62b1ea/osu.Game/Rulesets/Edit/HitObjectComposer.cs#L537

Some plus questions:

1. do we need this forced first object selection?
https://github.com/RatinFX/osu/blob/aa87e0a44d469bbdfb9abe9bca13ea0323eda774/osu.Game/Rulesets/Edit/EditorTimestampParser.cs#L55-L68

2. do we care about these types of unusual cases?
https://github.com/RatinFX/osu/blob/81caa854e6a7ba3251658f1cd174bf6bad62b1ea/osu.Game.Rulesets.Mania.Tests/Editor/TestSceneOpenEditorTimestampInMania.cs#L98-L99

bdach commented on osu.Game/Rulesets/Edit/EditorTimestampParser.cs:
```diff
@@ -0,0 +1,73 @@
+    {
+        // 00:00:000 (1,2,3) - test
+        // regex from https://github.com/ppy/osu-web/blob/651a9bac2b60d031edd7e33b8073a469bf11edaa/resources/assets/coffee/_classes/beatmap-discussion-helper.coffee#L10
+        public static readonly Regex TIME_REGEX = new Regex(@"\b(((\d{2,}):([0-5]\d)[:.](\d{3}))(\s\((?:\d+[,|])*\d+\))?)");
```
This regex is still doing things like looking for digits, commas, or pipes. All of that should be down to the ruleset.

RatinFX replied:
Do we change that part to `[^)]` so it accepts _anything_ between the parentheses?
- `\b(((\d{2,}):([0-5]\d)[:.](\d{3}))(\s\((?:\d+[,|])*\d+\))?)`
- `\b(((\d{2,}):([0-5]\d)[:.](\d{3}))(\s\([^)]+\))?)`

Naturally, this includes the alphabet and other characters as well. It should be fine in terms of compatibility with Stable as it responds with an "Invalid format!" notification whenever you try to click on an unrecognized editor link.

If this is okay, we might also need to change this same regex in `osu-web` at:
https://github.com/ppy/osu-web/blob/3b1698639244cfdaf0b41c68bfd651ea729ec2e3/resources/js/utils/beatmapset-discussion-helper.ts#L78

bdach commented on osu.Game/Rulesets/Edit/EditorTimestampParser.cs:
```diff
@@ -0,0 +1,73 @@
+
+            // Stable behavior
+            // - always selects the next closest object when `objectsGroup` only has one (combo) item
+            if (objectsToSelect.Length != 1 || objectsGroup.Contains('|'))
```
Similarly, the pipe character should not be getting considered at this level of abstraction.

RatinFX replied:
Refactored in [`4e1e197`](https://github.com/ppy/osu/pull/25371/commits/4e1e19728cb8ff9e11f18ab9c0e8635c2cc2ba9a)

bdach commented on osu.Game/Rulesets/Edit/HitObjectComposer.cs:
```diff
@@ -528,6 +528,19 @@ protected HitObjectComposer(Ruleset ruleset)
+        /// <returns>Whether a HitObject should be selected or not</returns>
+        public virtual bool HandleHitObjectSelection(HitObject hitObject, string objectInfo) => false;
+
+        /// <summary>
+        /// A character that separates the selection in <see cref="ConvertSelectionToString"/>
+        /// </summary>
+        public virtual char ObjectSeparator => ',';
```
This virtual member is much too specific and should not exist.

RatinFX replied:
Removed in [`4e1e197`](https://github.com/ppy/osu/pull/25371/commits/4e1e19728cb8ff9e11f18ab9c0e8635c2cc2ba9a)

bdach commented on osu.Game/Rulesets/Edit/HitObjectComposer.cs:
```diff
@@ -528,6 +528,19 @@ protected HitObjectComposer(Ruleset ruleset)
 
         public virtual string ConvertSelectionToString() => string.Empty;
 
+        /// <summary>
+        /// The custom logic that decides whether a HitObject should be selected when clicking an editor timestamp link
+        /// </summary>
+        /// <param name="hitObject">The hitObject being checked</param>
+        /// <param name="objectInfo">A single hitObject's information created with <see cref="ConvertSelectionToString"/></param>
+        /// <returns>Whether a HitObject should be selected or not</returns>
+        public virtual bool HandleHitObjectSelection(HitObject hitObject, string objectInfo) => false;
```
This also looks off. Why does this method take in a `HitObject` _and_ some string? The `HitObject` should unambiguously indicate what to select.

RatinFX replied:
Refactored in [`4e1e197`](https://github.com/ppy/osu/pull/25371/commits/4e1e19728cb8ff9e11f18ab9c0e8635c2cc2ba9a)

bdach commented:
To augment the above review, the way I see this being handled is:

- The timestamp is parsed generically regardless of ruleset.
- Anything, _anything_ placed between the brackets after, is interpreted _by_ the ruleset _in its entirety_.  So no weird half and half "well the separator is overridable but still hardcoded in the generic code" stuff. Ruleset gets `1,2,3`, or `1|3|5`, and sets up the selection as it sees fit. The API here should be a single method, something along the shape of `void SelectHitObjects(double timestamp, string objectDescription)`.

RatinFX committed:
Refactor HitObject selection in Composer

RatinFX committed:
Update Test cases

RatinFX committed:
Changed TIME_REGEX to accept everything in the parentheses
- changed osu-web link to the current value

bdach commented on osu.Game.Rulesets.Mania/Edit/ManiaHitObjectComposer.cs:
```diff
@@ -50,22 +49,44 @@ protected override ComposeBlueprintContainer CreateBlueprintContainer()
-            return Math.Abs(maniaHitObject.StartTime - timeValue) < 0.5
-                   && Math.Abs(maniaHitObject.Column - columnValue) < 0.5;
+            if (!double.TryParse(split[0], out double time) || !int.TryParse(split[1], out int column))
+                return false;
```
The fact this validation is done here is kind of wasteful because it means it will be done separately for every object at or after the first timestamp. Invalid object references should be culled before attempting to match them to objects.

Same thing applies to the osu! implementation.

RatinFX committed:
Do Split and Parse before checking HitObjects

RatinFX committed:
Updated comments, renamed method

bdach committed:
Improve xmldoc

bdach committed:
Simplify parsing code

Less methods, less smeared around logic, saner data types.

bdach committed:
Remove unnecessary guard

Setting a bindable's value to something if that value is already there
is a no-op (doesn't trigger bindings / callbacks).

bdach committed:
Rectify selection keep-alive logic

bdach committed:
Rename and document `SelectFromTimestamp`

bdach committed:
Fix tests

bdach committed:
Compile regexes

bdach committed:
Rename variable

bdach committed:
More test cleanup

bdach commented:
I've done a round of fix-ups here. Looks fine at this point, but due to the extent of changes made I'll probably request a double-check.

peppy committed:
Reorder and simplify private helper methods

peppy committed:
Adjust localisablel strings and keys

peppy committed:
Use `SimpleErrorNotification` for error display
